crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3362,6 +3364,51 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn fstring_expression() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            x = 0
+            y = str()
+
+            a = f'hello'
+            b = f'h {x}'
+            c = 'one ' f'single ' f'literal'
+            d = 'first ' f'second({b})' f' third'
+            e = f'-{y}-'
+            f = f'-{y}-' f'--' '--'
+            ",
+        )?;
+
+        assert_public_ty(&db, "src/a.py", "a", "Literal[\"hello\"]");
+        assert_public_ty(&db, "src/a.py", "b", "Literal[\"h 0\"]");
+        assert_public_ty(&db, "src/a.py", "c", "Literal[\"one single literal\"]");
+        assert_public_ty(&db, "src/a.py", "d", "Literal[\"first second(h 0) third\"]");
+        assert_public_ty(&db, "src/a.py", "e", "str");
+        assert_public_ty(&db, "src/a.py", "f", "str");
+
+        // More realistic use-case for inferring literals inside f-strings
+        db.write_dedented(
+            "src/endpoint.py",
+            "
+            BASE_URL = 'https://httpbin.org'
+            VERSION = 'v1'
+            endpoint = f'{BASE_URL}/{VERSION}/post'
+            ",
+        )?;
+        assert_public_ty(
+            &db,
+            "src/endpoint.py",
+            "endpoint",
+            "Literal[\"https://httpbin.org/v1/post\"]",
+        );
```

Comment:
This doesn't look necessary to me; it doesn't test anything that isn't already tested above. I don't think tests necessarily need to demonstrate a realistic-looking use case. If you want to make the above test look more realistic, that's fine, but IMO it's not worth writing and inferring types in another file.

If we did have a reason to keep this, it should be a separate test method.

Reply:
I think I mainly had it to make sure my comment in the PR's body was accurate, and committed it by mistake. Good catch :)